runtime/src/crowdfund.rs

```diff
@@ -0,0 +1,539 @@
+
+	/// What to do with funds that were not withdrawn.
+	type OrphanedFunds: OnUnbalanced<NegativeImbalanceOf<Self>>;
+}
+
+pub type FundIndex = u32;
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct FundInfo<AccountId, Balance, Hash, BlockNumber, ParaId> {
+	/// The parachain that this fund has funded, if there is one. As long as this is `Some`, then
+	/// the funds may not be withdrawn and the fund cannot be dissolved.
+	parachain: Option<ParaId>,
+	/// The owning account who placed the deposit.
+	owner: AccountId,
+	/// The amount of deposit placed.
+	deposit: Balance,
+	/// The total amount raised.
+	raised: Balance,
+	/// Block number after which the funding must have succeeded. If not successful at this number
+	/// the everyone may withdraw their funds.
+	end: BlockNumber,
+	/// A hard-cap on the amount that may be contributed.
+	cap: Balance,
+	/// The most recent block that this had a contribution. Determines if we make a bid or not.
+	/// If this is `None`, then the last contribution was made outside of the ending period.
+	last_contribution: Option<BlockNumber>,
+	/// First slot in range to bid on; it's actually a LeasePeriod, but that's the same type as
+	/// BlockNumber.
+	first_slot: BlockNumber,
+	/// Last slot in range to bid on; it's actually a LeasePeriod, but that's the same type as
+	/// BlockNumber.
+	last_slot: BlockNumber,
+	/// The deployment data associated with this fund, if any. Once set it may not be reset. First
+	/// is the code hash, second is the initial head data.
+	deploy_data: Option<(Hash, Vec<u8>)>,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		/// Info on all of the funds.
+		Funds get(funds):
+			map FundIndex => Option<FundInfo<T::AccountId, BalanceOf<T>, T::Hash, T::BlockNumber, ParaIdOf<T>>>;
+
+		/// The total number of funds that have so far been allocated.
+		FundCount get(fund_count): FundIndex;
+
+		/// The funds that have had additional contributions during the last block. This is used
+		/// in order to determine which funds should submit updated bids.
+		NewRaise: Vec<FundIndex>;
+
+		/// True if the fund was ending at the last block.
+		WasEnding: bool;
+	}
+}
+
+decl_event!(
+	pub enum Event<T>
+		where
+	<T as system::Trait>::AccountId,
+	Balance = BalanceOf<T>,
+	ParaId = ParaIdOf<T>,
+	{
+		Created(FundIndex),
+		Contributed(AccountId, FundIndex, Balance),
+		Withdrew(AccountId, FundIndex, Balance),
+		Retiring(FundIndex),
+		Dissolved(FundIndex),
+		DeployDataFixed(FundIndex),
+		Onboarded(FundIndex, ParaId),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+		
+		/// Create a new crowdfunding campaign for a parachain slot deposit.
+		#[weight = TransactionWeight::Basic(100_000, 10)]
+		fn create(
+			origin,
+			#[compact] end: T::BlockNumber,
+			#[compact] cap: BalanceOf<T>,
+			#[compact] first_slot: T::BlockNumber,
```

Comment:
Missing check that `first_slot` is greater than `end`, but to really make sure `handle_bid` is infallible, we need to check that `first_slot` is greater than `first_lease_period` so that `new_bounded` in slot range will not fail.

Seems like we should call `new_bounded` somewhere in `create` to make sure that it will not fail later.

Reply:
I suggest that we do not let the user input `end`, and rather we automatically set end to:

```
// Last possible block of an auction
let end = early_end + T::EndingPeriod::get();
```

for the current auction. In addition, we will check that the auction `is_in_progress` for things like contributions.

Reply:
I have done this in `Automatically assign end for crowdfund`, which we can revert if you don't like the idea.

Reply:
the current slot might change before the auction succeeds, so `handle_bid` can always be fallible at present. so far the assumption is that nobody would want to bid on an auction that they couldn't use. maybe that's invalid. but calling `new_bounded` here won't help other than checking that `first_slot < last_slot <= first_slot + 3`.